### DISCUSSION TITLE
Coexistence Zigbee2MQTT  with other MQTT sources using HA autodiscovery

### DISCUSSION THREAD

**Discussion by ratermir** [Koenkk/zigbee2mqtt#22407] [General]
I am already running HA with Zigbee2MQTT with several sensors. It works fine. 

Also I am trying to develop tool, that will convert data from weather station to JSON and post them to MQTT queue to utilize  "autodiscovery" mechanism of HA. It uses (by default) topic "homeassistant/sensor/..." to share sensor config and separate topics for posting measure data. 

Everything works fine in the isolated environment, but whe I deploy it to real HA, where already runs Zigbee2MQTT, measure data is published correctly, but the "discovery" messages - published (by default) under topic "homeassistant/sensor" didn't apear. 

I found that everything works if the Zigbee2MQTT is stopped, but after start it, the "autodiscovery" topics disapeared again. So it is clear  that Zigbee2MQTT cleans all content of these (sub)topics. 

So my question is obvious: How can I share singhe autodiscovery topic among Zigbee2MQTT and other tools, that wanted to publis autodiscovery messages ? As I know, only one such topic can be configured in HA (the solutionwould also be if HA enabled multiple autodiscovery topics). Can I specify that Zigbee2MQTT don't touch - for example - topics with certin prefix? Or to disable such "cleaning" completly?

**Comment by francisp2** [comment]
Strange. My "homeassistant" tree is filled with autodiscovery topics from Zigbee2MQTT, Tasmota, OpenMQTTGateway and self-created topics.

**Comment by ratermir** [comment]
Maybe I am doing something wrong.  If I understood it well, the "discovery" topic has the fixed structure, it start with (configurable) topic_base, then the device class, device ID,  value type and config, so it looks like this: 
`my-discovery-topic/sensor/weather-station-in/temperature/config` 
`my-discovery-topic/sensor/weather-station-in/pressure/config` 
`my-discovery-topic/sensor/weather-station-out/temperature/config` 
etc ... for every measutment quantity provided by the station. The value of the topic is JSON configuration. So I must share the 
"`my-discovery-topic/sensor/...`" topics between Zigbee2MQTT and the weather station. 

If I follow this structure, everything works well in the isolated environment. But together with the Z2M, it behaves as described above. 
I.e. measurment data is preserved, but content of the weather station discovery topics - i.e. `my-discovery-topic/sensor/weather-station...` - is not seen in MQTT at all. There is only set of topics generated by Zigbee2MQTT for sensors conncted to it.

**Reply by francisp2** [reply]
![afbeelding](https://github.com/Koenkk/zigbee2mqtt/assets/52328611/c8282cb4-7770-4eba-9444-72124bd41cc6)

**Comment by ratermir** [comment]
Sure, this is what I'd expected. But in my case it behaves strangely.  In the log of  Z2M I found message as follow: 

``` 
[2024-05-06 16:42:05] debug:    z2m: Received MQTT message on 'zmh-lip_discovery/sensor/weather-lip-in/temperature/config' with data '
{
    "availability": [{
            "topic": "zmh-lip_data/weather-lip_in/availability",
            "value_template": "{{ value_json.state }}"
    }],
    "availability_mode": "all",
    "device": {
        "identifiers": "ratewate_weatherlipGARNI01",
        "manufacturer": "Garni",
        "name": "weather-lip",
        "model": "Garni weather station"
    },
    "device_class": "temperature",
    "enabled_by_default": true,
    "origin": {
        "name": "Ratewate",
        "sw": "0.0.1",
        "url": "https://github.com/ratermir/ratewate"
    },
    "name": "temperature in",
    "object_id": "weather-lip-temperature-in_temperature",
    "unique_id": "weatherlipGARNI01_temperature_in",
    "state_class": "measurement",
    "state_topic": "zmh-lip_data/weather-lip_in",
    "unit_of_measurement": "°C",
    "value_template": "{{ value_json.temperature }}"
}
'
[2024-05-06 16:42:05] debug:    z2m: Clearing Home Assistant config 'zmh-lip_discovery/sensor/weather-lip-in/temperature/config'
[2024-05-06 16:42:05] debug:    z2m: MQTT publish: topic 'zmh-lip_discovery/sensor/weather-lip-in/temperature/config', payload 'null'
```
The last logged message is the cause, but why it is pulished ?

This sequence repeats for every Non-Z2M sensor ...

**Reply by ratermir** [reply]
The cause is that Z2M really deletes all topics, that occur in it's  "availability" structure under its "base" topic and are not known directly to Z2M. 
Since I wanted to keep "root" topic structure simple, I wanted to share discovery and data topics with Z2M and all other devices. But it is not possible due above. The solution was move data topics (that contain also availability topics) under different root.
